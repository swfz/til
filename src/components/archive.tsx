import { Link, useStaticQuery, graphql } from "gatsby"
import isNil from "lodash/isNil"
import React from "react"

import { Archives, ArchiveMonth, ArchiveEdges } from "../@types"

type Props = {
  data: Queries.ArchiveQueryQuery
}

const Archive = () => {
  const data: Queries.ArchiveQueryQuery = useStaticQuery(graphql`
    query ArchiveQuery {
      allMarkdownRemark(limit: 2000) {
        edges {
          node {
            frontmatter {
              date
            }
          }
        }
      }
    }
  `)
  return <ArchiveList data={data}></ArchiveList>
}

export const ArchiveList = (props: Props) => {
  const archives = props.data.allMarkdownRemark.edges.reduce((acc, cur) => {
    if (isNil(cur) || isNil(cur.node)) {
      return acc
    }

    const date = cur?.node?.frontmatter?.date || ""
    const [year, month] = date.split("-")

    const ym = acc?.[year]?.[month] ? [...acc[year][month], cur] : [cur]

    const updateValue = { ...acc[year], [month]: ym }

    return { ...acc, [year]: updateValue }
  }, {} as Archives)

  const descFn = (a: [string, ArchiveMonth | ArchiveEdges], b: [string, ArchiveMonth | ArchiveEdges]) => {
    if (a[0] > b[0]) {
      return -1
    } else {
      return 1
    }
  }

  const isOpen = (year: number) => {
    const nowYear = new Date().getFullYear()

    return nowYear === year
  }

  return (
    <>
      <p>Archive</p>
      {Object.entries(archives)
        .sort(descFn)
        .map(([year, items]) => (
          <details open={isOpen(parseInt(year))} key={year}>
            <summary>
              <Link aria-label={"year-link"} to={`/archives/${year}`}>
                {year}({Object.entries(items).reduce((acc, [, v]) => acc + v.length, 0)})
              </Link>
            </summary>
            <ul>
              {Object.entries(items)
                .sort(descFn)
                .map(([month, items]) => (
                  <li aria-label={"month-link"} key={`${year}-${month}`}>
                    <Link to={`/archives/${year}/${month}`}>
                      <p>
                        &nbsp;&nbsp;&nbsp;{year}-{month}({items.length})
                      </p>
                    </Link>
                  </li>
                ))}
            </ul>
          </details>
        ))}
    </>
  )
}

export default Archive
