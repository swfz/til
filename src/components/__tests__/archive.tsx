import React from "react"
import "@testing-library/jest-dom/extend-expect"
import { render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import { Archive } from "../archive"
import renderer from "react-test-renderer"

describe("Archive", () => {
  const user = userEvent.setup()
  const data = {
    allMarkdownRemark: {
      edges: [
        { node: { frontmatter: { date: "2021-01-11" } } },
        { node: { frontmatter: { date: "2021-01-20" } } },
        { node: { frontmatter: { date: "2021-01-25" } } },
        { node: { frontmatter: { date: "2020-12-28" } } },
        { node: { frontmatter: { date: "2020-12-20" } } },
        { node: { frontmatter: { date: "2020-12-15" } } },
        { node: { frontmatter: { date: "2020-11-10" } } },
      ],
    },
  }
  it("renders correctly", async () => {
    const { getAllByText, getAllByLabelText } = render(
      <Archive data={data}></Archive>
    )

    // 記事件数と順番があっているか
    const yearLinkList = getAllByLabelText("year-link")
    expect(yearLinkList).toHaveLength(2)
    expect(yearLinkList[0]).toHaveTextContent("2021(3)")
    expect(yearLinkList[1]).toHaveTextContent("2020(4)")

    expect(yearLinkList[0]).toBeVisible()
    expect(yearLinkList[1]).toBeVisible()

    // 初回レンダリング時にリストが展開された状態で表示されているか
    const monthLInkList = getAllByLabelText("month-link")
    expect(monthLInkList).toHaveLength(3)
    expect(monthLInkList[0]).toHaveTextContent("2021-01(3)")
    expect(monthLInkList[1]).toHaveTextContent("2020-12(3)")
    expect(monthLInkList[2]).toHaveTextContent("2020-11(1)")

    expect(monthLInkList[0]).toBeVisible()
    expect(monthLInkList[1]).toBeVisible()
    expect(monthLInkList[2]).toBeVisible()

    // +ボタンクリック後にリストが閉じた状態になっているか
    const plusText2021 = getAllByText(/\+/i)[0]
    await user.click(plusText2021)

    expect(monthLInkList[0]).not.toBeVisible()
    expect(monthLInkList[1]).toBeVisible()
    expect(monthLInkList[2]).toBeVisible()

    const plusText2020 = getAllByText(/\+/i)[1]
    await user.click(plusText2020)

    expect(monthLInkList[0]).not.toBeVisible()
    expect(monthLInkList[1]).not.toBeVisible()
    expect(monthLInkList[2]).not.toBeVisible()
  })
  it("snapshot", () => {
    const tree = renderer.create(<Archive data={data}></Archive>).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
