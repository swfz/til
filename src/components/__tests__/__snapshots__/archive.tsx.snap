// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Archive snapshot 1`] = `
[
  <p>
    Archive
  </p>,
  <details
    open={true}
  >
    <summary>
      +
      <a
        aria-label="year-link"
        href="/archives/2021"
      >
        2021
        (
        3
        )
      </a>
    </summary>
    <ul>
      <li
        aria-label="month-link"
      >
        <a
          href="/archives/2021/01"
        >
          <p>
               
            2021
            -
            01
            (
            3
            )
          </p>
        </a>
      </li>
    </ul>
  </details>,
  <details
    open={false}
  >
    <summary>
      +
      <a
        aria-label="year-link"
        href="/archives/2020"
      >
        2020
        (
        4
        )
      </a>
    </summary>
    <ul>
      <li
        aria-label="month-link"
      >
        <a
          href="/archives/2020/12"
        >
          <p>
               
            2020
            -
            12
            (
            3
            )
          </p>
        </a>
      </li>
      <li
        aria-label="month-link"
      >
        <a
          href="/archives/2020/11"
        >
          <p>
               
            2020
            -
            11
            (
            1
            )
          </p>
        </a>
      </li>
    </ul>
  </details>,
]
`;
